File: compile/factorial.c
#include <stdio.h>
#include <stdlib.h>
#include <gmp.h> // Include the GMP header

// Function to calculate factorial of a number using GMP
void factorial(int n)
{
    mpz_t result;     // Declare a GMP integer
    mpz_init(result); // Initialize the integer

    mpz_fac_ui(result, n); // Calculate factorial

    // Print the factorial result
    gmp_printf("%Zd", result);

    mpz_clear(result); // Free the memory used by the GMP integer
}

int main(int argc, char *argv[])
{
    if (argc != 2)
    {
        printf("Usage: %s <number>\n", argv[0]);
        return 1;
    }
    int n = atoi(argv[1]);
    factorial(n); // Call the factorial function
    return 0;
}

File: compile/factorial.php
<?php

declare(strict_types=1);

function factorial(int $number): string
{
    $binaryPath = __DIR__ . '/factorial';

    if (!file_exists($binaryPath)) {
        throw new Exception('Library not found');
    }

    $output = shell_exec("$binaryPath $number");
    if ($output === false) {
        throw new Exception('Failed to execute');
    }

    return $output;
}

const REQUIRED_ARGUMENTS = 2;

if ($argc !== REQUIRED_ARGUMENTS) {
    echo "Usage: php factorial.php <number>\n";
    exit(1);
}

$number = (int)$argv[1];

$result = factorial($number);

echo "Factorial of $number is: $result\n";
